---
title: "snowcrabs_transcriptome"
author: "Courtney Skalley"
date: "2023-05-18"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Download Software

```{r}
library(tidyverse)
```

# check that blast has been downloaded
```{bash}
ls /home/shared/
```


# make blast database

```{bash, blast database}
cd ../data
curl -O https://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/complete/uniprot_sprot.fasta.gz
mv uniprot_sprot.fasta.gz uniprot_sprot_r2023_01.fasta.gz
gunzip -k uniprot_sprot_r2023_01.fasta.gz
ls ../data
```

```{bash}
/home/shared/ncbi-blast-2.11.0+/bin/makeblastdb \
-in ../data/uniprot_sprot_r2023_01.fasta \
-dbtype prot \
-out ../blastdb/uniprot_sprot_r2023_01
```

# see all available query sequences

```{bash}
ls /home/shared/8TB_HDD_01/snow_crab/5010/
```

#PERMISSION DENIED unzip the first fasta file
```{bash}
gunzip /home/shared/8TB_HDD_01/snow_crab/5010/5010_10_S10_L003_R1_001.fastq.gz
mv /home/shared/8TB_HDD_01/snow_crab/5010/5010_10_S10_L003_R1_001.fastq ../data/5010_10_S10_L003_R1_001.fastq
```

# examaine 
```{bash}
head /home/shared/8TB_HDD_01/snow_crab/5010/5010_10_S10_L003_R1_001.fastq.gz

# if moved correctly: ../data/5010_10_S10_L003_R1_001.fastq

echo "How many sequences are there?"
grep -c ">" /home/shared/8TB_HDD_01/snow_crab/5010/5010_10_S10_L003_R1_001.fastq.gz

# if moved correctly: grep -c ">" ../data/5010_10_S10_L003_R1_001.fastq
#examine the first fasta file

```{bash}
head /home/shared/8TB_HDD_01/snow_crab/5010/5010_10_S10_L003_R1_001.fastq.gz
echo "How many sequences are there?"
grep -c ">" /home/shared/8TB_HDD_01/snow_crab/5010/5010_10_S10_L003_R1_001.fastq.gz
```

# run blast

```{bash}
/home/shared/ncbi-blast-2.11.0+/bin/blastx \
-query /home/shared/8TB_HDD_01/snow_crab/5010/5010_10_S10_L003_R1_001.fastq.gz \
-db ../blastdb/uniprot_sprot_r2023_01 \
-out ../output/c_opilio_uniprot_blastx.tab \
-evalue 1E-20 \
-num_threads 20 \
-max_target_seqs 1 \
-outfmt 6
```



```{bash}
head -2 ../output/c_opilio_uniprot_blastx.tab
wc -l ../output/c_opilio_uniprot_blastx.tab
head ../data/Ab_4denovo_CLC6_a.fa
echo "How many sequences are there?"
grep -c ">" ../data/Ab_4denovo_CLC6_a.fa
```



```{bash}
~/applications/ncbi-blast-2.13.0+/bin/blastx \
-query ../data/Ab_4denovo_CLC6_a.fa \
-db ../blastdb/uniprot_sprot_r2023_01 \
-out ../output/Ab_4-uniprot_blastx.tab \
-evalue 1E-20 \
-num_threads 20 \
-max_target_seqs 1 \
-outfmt 6
```



```{bash}
head -2 ../output/Ab_4-uniprot_blastx.tab
wc -l ../output/Ab_4-uniprot_blastx.tab
```

&&& this puts table in code directory

```{bash}
curl -o "uniprot_table_r2023_01.tab" -H "Accept: text/plain; format=tsv" "https://gannet.fish.washington.edu/seashell/snaps/uniprot_table_r2023_01.tab"
```


```{bash}
head -2 ../output/Ab_4-uniprot_blastx.tab
wc -l ../output/Ab_4-uniprot_blastx.tab
```

```{bash}
tr '|' '\t' < ../output/Ab_4-uniprot_blastx.tab | head -2
```
```{bash}
tr '|' '\t' < ../output/Ab_4-uniprot_blastx.tab \
> ../output/Ab_4-uniprot_blastx_sep.tab
```

#this chunk works but is not consistent with the assignement instructions (it uses a .fasta file instead 
# of a .tab)

&&& code does not work as table is in code directory (see above)

```{bash}
head -2 ../data/uniprot_table_r2023_01.tab
wc -l ../data/uniprot_table_r2023_01.tab
```

```{r}
bltabl <- read.csv("../output/Ab_4-uniprot_blastx_sep.tab", sep = '\t', header = FALSE)
```


&&& I altered below to work

```{r}
spgo <- read.csv("../code/uniprot_table_r2023_01.tab", sep = '\t', header = TRUE)
```

```{r}
str(spgo)
```


```{r}
left_join(bltabl, spgo,  by = c("V3" = "Entry")) %>%
  select(V1, V3, V13, Protein.names, Organism, Gene.Ontology..biological.process., Gene.Ontology.IDs) %>% mutate(V1 = str_replace_all(V1, 
            pattern = "solid0078_20110412_FRAG_BC_WHITE_WHITE_F3_QV_SE_trimmed", replacement = "Ab")) %>%
  write_delim("../output/blast_annot_go.tab", delim = '\t')
```

```{r}
annot_tab <- read.csv("../output/blast_annot_go.tab", sep = '\t', header = TRUE)
```

```{r}
str(annot_tab)
```